File: python/1d-dynamic-programming/decode-ways.py
class Solution:
    def numDecodings(self, s: str) -> int:
        cache = {}

        def dfs(index):
            if index == len(s):
                return 1
            if s[index] == "0":
                return 0
            if index in cache:
                return cache[index]

            cache[index] = dfs(index + 1)
            if index < len(s) - 1 and (s[index] == "1" or (s[index] == "2" and int(s[index + 1]) < 7)):
                cache[index] += dfs(index + 2)

            return cache[index]

        return dfs(0)

File: c++/1d-dynamic-programming/decode-ways.cc
class Solution {
public:
    int numDecodings(string s) {
        std::unordered_map<int, int> cache;
        return decode(0, s, cache);
    }

private:
    int decode(int i, std::string& s, std::unordered_map<int, int>& cache) {
        if (i == s.size()) {
            return 1;
        }

        if (s[i] == '0') {
            return 0;
        }

        if (cache.contains(i)) {
            return cache[i];
        }

        cache[i] = 0;
        if (i < s.size() - 1 && (s[i] == '1' || (s[i] == '2' && s[i+1] - '0' < 7))) {
            cache[i] += decode(i + 2, s, cache);
        }
        cache[i] += decode(i + 1, s, cache);

        return cache[i];
    }
};
